File: js/plugins/google-genai/src/googleai/deep-research.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ActionMetadata, modelActionMetadata, z } from 'genkit';
import {
  BackgroundModelAction,
  ModelInfo,
  ModelReference,
  modelRef,
} from 'genkit/model';
import { backgroundModel as pluginBackgroundModel } from 'genkit/plugin';
import {
  ensureToolIds,
  fromInteraction,
  toInteractionTool,
  toInteractionTurn,
} from '../common/interaction-converters.js';
import {
  InteractionTool,
  ResponseModality,
} from '../common/interaction-types.js';
import { cleanSchema, isKnownKey } from '../common/utils.js';
import {
  cancelInteraction,
  createInteraction,
  getInteraction,
} from './client.js';
import {
  ClientOptions,
  CreateInteractionRequest,
  GoogleAIPluginOptions,
  Model,
} from './types.js';
import {
  calculateApiKey,
  checkApiKey,
  checkModelName,
  extractVersion,
  modelName,
} from './utils.js';

export const DeepResearchConfigSchema = z
  .object({
    apiKey: z
      .string()
      .describe('Override the API key provided at plugin initialization.')
      .optional(),
    baseUrl: z
      .string()
      .describe(
        'Overrides the plugin-configured or default baseUrl, if specified.'
      )
      .optional(),
    apiVersion: z
      .string()
      .describe(
        'Overrides the plugin-configured or default apiVersion, if specified.'
      )
      .optional(),
    thinkingSummaries: z
      .enum(['AUTO', 'NONE'])
      .describe('Whether to include thought summaries in the response.')
      .optional(),
    previousInteractionId: z
      .string()
      .describe('The ID of the previous interaction, if any.')
      .optional(),
    store: z
      .boolean()
      .describe(
        'Whether to store the response and request for later retrieval.'
      )
      .optional(),
    responseModalities: z
      .array(z.enum(['TEXT', 'IMAGE', 'AUDIO']))
      .describe('The modalities to be used in response.')
      .optional(),
    visualization: z
      .enum(['AUTO', 'OFF', 'auto', 'off'])
      .describe('Whether to enable agent-generated charts and images.')
      .optional(),
    collaborativePlanning: z
      .boolean()
      .describe(
        'Whether to enable multi-turn plan review before research begins.'
      )
      .optional(),
    googleSearch: z
      .union([z.boolean(), z.object({}).passthrough()])
      .describe(
        'Search the public web. Enabled by default if no tools are provided.'
      )
      .optional(),
    urlContext: z
      .union([z.boolean(), z.object({}).passthrough()])
      .describe(
        'Read and summarize web page content. Enabled by default if no tools are provided.'
      )
      .optional(),
    codeExecution: z
      .union([z.boolean(), z.object({}).passthrough()])
      .describe(
        'Execute code to perform calculations and data analysis. Enabled by default if no tools are provided.'
      )
      .optional(),
    fileSearch: z
      .object({
        fileSearchStoreNames: z.array(z.string()),
      })
      .passthrough()
      .optional(),
    mcpServers: z
      .array(
        z
          .object({
            name: z.string().optional(),
            url: z.string(),
            headers: z.record(z.string()).optional(),
            allowedTools: z.array(z.string()).optional(),
          })
          .passthrough()
      )
      .optional(),
  })
  .passthrough();
export type DeepResearchConfigSchemaType = typeof DeepResearchConfigSchema;
export type DeepResearchConfig = z.infer<DeepResearchConfigSchemaType>;

// This contains all the DeepResearch config schema types
type ConfigSchemaType = DeepResearchConfigSchemaType;

function commonRef(
  name: string,
  info?: ModelInfo,
  configSchema: ConfigSchemaType = DeepResearchConfigSchema
): ModelReference<ConfigSchemaType> {
  return modelRef({
    name: `googleai/${name}`,
    configSchema,
    info:
      info ??
      ({
        supports: {
          multiturn: true,
          media: false,
          tools: false,
          toolChoice: false,
          systemRole: false,
          output: ['text'],
          longRunning: true,
        },
      } as ModelInfo),
  });
}

const GENERIC_MODEL = commonRef('deep-research');

const ADVANCED_DEEP_RESEARCH_INFO = {
  supports: {
    multiturn: true,
    media: true,
    tools: true,
    toolChoice: false,
    systemRole: false,
    output: ['text', 'media'],
    longRunning: true,
  },
} as ModelInfo;

const KNOWN_MODELS = {
  'deep-research-pro-preview-12-2025': commonRef(
    'deep-research-pro-preview-12-2025'
  ),
  'deep-research-preview-04-2026': commonRef(
    'deep-research-preview-04-2026',
    ADVANCED_DEEP_RESEARCH_INFO
  ),
  'deep-research-max-preview-04-2026': commonRef(
    'deep-research-max-preview-04-2026',
    ADVANCED_DEEP_RESEARCH_INFO
  ),
} as const;
export type KnownModels = keyof typeof KNOWN_MODELS; // For autocomplete

export type DeepResearchModelName = `deep-research-${string}`;
export function isDeepResearchModelName(
  value?: string
): value is DeepResearchModelName {
  return !!value?.startsWith('deep-research-');
}

export function model(
  version: string,
  config: DeepResearchConfig = {}
): ModelReference<ConfigSchemaType> {
  const name = checkModelName(version);

  if (isKnownKey(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }

  return modelRef({
    name: `googleai/${name}`,
    config,
    configSchema: DeepResearchConfigSchema,
    info: { ...GENERIC_MODEL.info },
  });
}

export function listActions(models: Model[]): ActionMetadata[] {
  return (
    models
      .filter((m) => isDeepResearchModelName(modelName(m.name)))
      // Filter out deprecated
      .filter((m) => !m.description || !m.description.includes('deprecated'))
      .map((m) => {
        const ref = model(m.name);
        return modelActionMetadata({
          name: ref.name,
          info: ref.info,
          configSchema: ref.configSchema,
        });
      })
  );
}

export function listKnownModels(options?: GoogleAIPluginOptions) {
  return Object.keys(KNOWN_MODELS).map((name: string) =>
    defineModel(name, options)
  );
}

export function defineModel(
  name: string,
  pluginOptions?: GoogleAIPluginOptions
): BackgroundModelAction<DeepResearchConfigSchemaType> {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions: ClientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders,
  };

  return pluginBackgroundModel({
    name: ref.name,
    ...ref.info,
    configSchema: ref.configSchema,
    async start(request) {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        thinkingSummaries,
        visualization,
        collaborativePlanning,
        previousInteractionId,
        store,
        responseModalities,
        googleSearch,
        urlContext,
        codeExecution,
        fileSearch,
        mcpServers,
        ...rest
      } = request.config || {};
      const apiKey = calculateApiKey(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions: ClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion,
      };

      const messages = structuredClone(request.messages);
      // Deep Research does not support system instructions, so we map them to
      // user messages.
      for (const message of messages) {
        if (message.role === 'system') {
          message.role = 'user';
        }
      }

      const tools: InteractionTool[] = [];
      if (request.tools?.length) {
        for (const tool of request.tools) {
          tools.push(toInteractionTool(tool));
        }
      }

      if (googleSearch) {
        tools.push({
          type: 'google_search',
          ...(googleSearch === true ? {} : googleSearch),
        } as InteractionTool);
      }
      if (urlContext) {
        tools.push({
          type: 'url_context',
          ...(urlContext === true ? {} : urlContext),
        } as InteractionTool);
      }
      if (codeExecution) {
        tools.push({
          type: 'code_execution',
          ...(codeExecution === true ? {} : codeExecution),
        } as InteractionTool);
      }
      if (fileSearch) {
        tools.push({
          type: 'file_search',
          file_search_store_names: fileSearch.fileSearchStoreNames,
          ...fileSearch,
        } as InteractionTool);
      }
      if (mcpServers) {
        for (const mcpServer of mcpServers) {
          const { allowedTools, ...restMcp } = mcpServer;
          tools.push({
            type: 'mcp_server',
            ...restMcp,
            ...(allowedTools ? { allowed_tools: allowedTools } : {}),
          } as InteractionTool);
        }
      }

      let responseMimeType = request.output?.contentType;
      let responseFormat: any = undefined;
      if (
        request.output?.format === 'json' ||
        request.output?.contentType === 'application/json'
      ) {
        responseMimeType = 'application/json';
        if (request.output.schema) {
          responseFormat = cleanSchema(request.output.schema);
        }
      }

      let responseModalitiesConverted: ResponseModality[] | undefined;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase() as ResponseModality
        );
      }

      const req: CreateInteractionRequest = {
        agent: extractVersion(ref),
        input: ensureToolIds(messages).map(toInteractionTurn),
        background: true,
        agent_config: {
          type: 'deep-research',
          thinking_summaries: thinkingSummaries?.toLowerCase() as
            | 'auto'
            | 'none'
            | undefined,
          visualization: visualization?.toLowerCase() as
            | 'auto'
            | 'off'
            | undefined,
          collaborative_planning: collaborativePlanning,
        },
        previous_interaction_id: previousInteractionId,
        store,
        tools: tools.length ? tools : undefined,
        response_mime_type: responseMimeType,
        response_format: responseFormat,
        response_modalities: responseModalitiesConverted,
        ...rest,
      };

      const response = await createInteraction(apiKey, req, newClientOptions);

      return fromInteraction<ClientOptions>(response, newClientOptions);
    },
    async check(operation) {
      const storedOptions = operation.metadata?.clientOptions as
        | ClientOptions
        | undefined;
      const apiKey =
        storedOptions?.apiKey ||
        calculateApiKey(pluginOptions?.apiKey, undefined);

      const checkOptions: ClientOptions = {
        ...clientOptions,
        ...storedOptions,
      };

      const response = await getInteraction(apiKey, operation.id, checkOptions);

      return fromInteraction<ClientOptions>(response, checkOptions);
    },
    async cancel(operation) {
      const storedOptions = operation.metadata?.clientOptions as
        | ClientOptions
        | undefined;
      const apiKey =
        storedOptions?.apiKey ||
        calculateApiKey(pluginOptions?.apiKey, undefined);

      const cancelOptions: ClientOptions = {
        ...clientOptions,
        ...storedOptions,
      };

      const response = await cancelInteraction(
        apiKey,
        operation.id,
        cancelOptions
      );

      return fromInteraction<ClientOptions>(response, cancelOptions);
    },
  });
}

export const TEST_ONLY = { KNOWN_MODELS };

File: py/packages/genkit-google-genai/src/genkit_google_genai/models/deep_research.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Google AI Interactions Deep Research background model action."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, field_validator
from pydantic.alias_generators import to_camel

from genkit import ModelRequest
from genkit.model import BackgroundAction, ModelRef, Operation, model_ref
from genkit.plugin_api import Action, ActionKind, ActionRunContext, to_json_schema
from genkit_google_genai._interactions.client import (
    cancel_interaction,
    create_interaction,
    get_interaction,
)
from genkit_google_genai._interactions.converters import (
    clean_schema,
    ensure_tool_ids,
    from_interaction,
    split_system_instruction,
    to_interaction_steps,
    to_interaction_tool,
)
from genkit_google_genai._interactions.options import ClientOptions
from genkit_google_genai.models._secrets import reject_request_config_api_key
from genkit_google_genai.models.interactions_registry import deep_research_model_info
from genkit_google_genai.models.interactions_utils import (
    api_key_for_context,
    client_overrides_from_config,
    extract_version,
    lowercase_choice,
    lowercase_choice_list,
    partition_keys,
    remove_client_option_overrides,
    require_interaction_steps,
)

AGENT_CONFIG_KEYS = (
    'thinking_summaries',
    'visualization',
    'collaborative_planning',
)
TOOL_CONFIG_KEYS = (
    'google_search',
    'url_context',
    'code_execution',
    'file_search',
    'mcp_servers',
)
CREATE_OPTION_KEYS = (
    'previous_interaction_id',
    'store',
    'response_modalities',
)


class McpServerConfig(BaseModel):
    """MCP server configuration for Deep Research."""

    model_config = ConfigDict(extra='allow', populate_by_name=True, alias_generator=to_camel)
    name: str | None = None
    url: str | None = None
    headers: dict[str, str] | None = None
    allowed_tools: list[str] | None = None


class FileSearchConfig(BaseModel):
    """File search store configuration for Deep Research."""

    model_config = ConfigDict(extra='allow', populate_by_name=True, alias_generator=to_camel)
    file_search_store_names: list[str]


class DeepResearchConfig(BaseModel):
    """Deep Research model configuration."""

    model_config = ConfigDict(extra='allow', populate_by_name=True, alias_generator=to_camel)
    base_url: str | None = None
    api_version: str | None = None
    # Milliseconds — applied to the HTTP call, not the create body.
    timeout: float | None = None
    custom_headers: dict[str, str] | None = None
    thinking_summaries: Literal['auto', 'none'] | None = None
    previous_interaction_id: str | None = None
    store: bool | None = None
    response_modalities: list[Literal['text', 'image', 'audio']] | None = None
    visualization: Literal['auto', 'off'] | None = None
    collaborative_planning: bool | None = None
    google_search: bool | None = None
    url_context: bool | None = None
    code_execution: bool | None = None
    file_search: FileSearchConfig | None = None
    mcp_servers: list[McpServerConfig] | None = None

    @field_validator('thinking_summaries', 'visualization', mode='before')
    @classmethod
    def fold_choice_case(cls, value: object) -> object:
        """Accept AUTO/NONE the same as auto/none."""
        return lowercase_choice(value)

    @field_validator('response_modalities', mode='before')
    @classmethod
    def fold_modalities_case(cls, value: object) -> object:
        """Accept TEXT/IMAGE/AUDIO the same as lowercase wire values."""
        return lowercase_choice_list(value)


def deep_research_model(version: str) -> ModelRef:
    """Return a ModelRef for a Deep Research version (namespaced or bare)."""
    clean = extract_version(version)
    return model_ref(
        name=clean,
        namespace='googleai',
        info=deep_research_model_info(clean),
        config_schema=DeepResearchConfig,
    )


def build_tools(request: ModelRequest[DeepResearchConfig], config: DeepResearchConfig) -> list[dict[str, Any]]:
    """Build Interactions API tool configurations for Deep Research."""
    tools: list[dict[str, Any]] = []
    if request.tools:
        tools.extend(dict(to_interaction_tool(tool_def)) for tool_def in request.tools)

    if config.google_search:
        tools.append({'type': 'google_search'})
    if config.url_context:
        tools.append({'type': 'url_context'})
    if config.code_execution:
        tools.append({'type': 'code_execution'})
    if config.file_search is not None:
        # Create body is snake_case. Dumping by alias would send
        # fileSearchStoreNames and the stores would never attach.
        tools.append({'type': 'file_search', **config.file_search.model_dump(exclude_none=True)})
    for mcp_server in config.mcp_servers or []:
        tools.append({'type': 'mcp_server', **mcp_server.model_dump(exclude_none=True)})

    return tools


def response_format_from_request(
    request: ModelRequest[DeepResearchConfig],
) -> dict[str, Any] | None:
    """Build response_format when the caller asked for JSON output."""
    if request.output_format != 'json' and request.output_content_type != 'application/json':
        return None
    response_format: dict[str, Any] = {'type': 'text', 'mime_type': 'application/json'}
    if request.output_schema:
        # output_schema is already a JSON Schema dict. Dumping a model
        # by alias here would rename properties the constraint expects.
        response_format['schema'] = clean_schema(request.output_schema)
    return response_format


def create_deep_research_background_action(
    target: str | ModelRef,
    *,
    plugin_api_key: str | None,
    client_options: ClientOptions,
) -> BackgroundAction:
    """Wire Deep Research start/check/cancel as Actions that re-read secrets."""
    name = target.name if isinstance(target, ModelRef) else target
    version = extract_version(name)
    info = deep_research_model_info(version)
    full_name = name if '/' in name else f'googleai/{name}'
    action_key = f'/background-model/{full_name}'

    def persist(operation: Operation) -> Operation:
        operation.action = action_key
        return operation

    async def start(request: ModelRequest[DeepResearchConfig], ctx: ActionRunContext) -> Operation:
        reject_request_config_api_key(request.config)
        config = request.config or DeepResearchConfig()
        api_key = api_key_for_context(ctx.context, plugin_api_key)
        options = client_options.merge(
            client_overrides_from_config(
                base_url=config.base_url,
                api_version=config.api_version,
                timeout=config.timeout,
                custom_headers=config.custom_headers,
            )
        )

        dumped = remove_client_option_overrides(config.model_dump(exclude_none=True))
        agent_fields, _tool_fields, create_options, passthrough = partition_keys(
            dumped,
            AGENT_CONFIG_KEYS,
            TOOL_CONFIG_KEYS,
            CREATE_OPTION_KEYS,
        )
        agent_config: dict[str, Any] = {'type': 'deep-research', **agent_fields}

        tools = build_tools(request, config)
        response_format = response_format_from_request(request)
        # Deep Research rejects system_instruction and asks for guidance in the
        # input prompt instead, so system turns become a leading user_input step.
        system_instruction, turns = split_system_instruction(request.messages or [])
        steps = to_interaction_steps(ensure_tool_ids(turns))
        if system_instruction:
            steps.insert(
                0,
                {
                    'type': 'user_input',
                    'content': [{'type': 'text', 'text': system_instruction}],
                },
            )
        require_interaction_steps(steps)
        create_kwargs: dict[str, Any] = {
            'agent': version,
            'input': steps,
            'background': True,
            'agent_config': agent_config,
            **create_options,
            **passthrough,
        }
        if tools:
            create_kwargs['tools'] = tools
        if response_format is not None:
            create_kwargs['response_format'] = response_format

        interaction = await create_interaction(api_key, create_kwargs, options)
        return persist(from_interaction(interaction))

    async def check(operation: Operation, ctx: ActionRunContext) -> Operation:
        call_config = ctx.context.get('config')
        options = client_options.merge(call_config if isinstance(call_config, dict) else None)
        api_key = api_key_for_context(ctx.context, plugin_api_key)
        interaction = await get_interaction(api_key, operation.id, options)
        return persist(from_interaction(interaction))

    async def cancel(operation: Operation, ctx: ActionRunContext) -> Operation:
        call_config = ctx.context.get('config')
        options = client_options.merge(call_config if isinstance(call_config, dict) else None)
        api_key = api_key_for_context(ctx.context, plugin_api_key)
        interaction = await cancel_interaction(api_key, operation.id, options)
        return persist(from_interaction(interaction))

    start_action = Action(
        kind=ActionKind.BACKGROUND_MODEL,
        name=full_name,
        fn=start,
        metadata={
            'model': {**info.model_dump(by_alias=True), 'customOptions': to_json_schema(DeepResearchConfig)},
            'type': 'background-model',
        },
    )
    check_action = Action(
        kind=ActionKind.CHECK_OPERATION,
        name=f'{full_name}/check',
        fn=check,
        metadata={'type': 'check-operation'},
    )
    cancel_action = Action(
        kind=ActionKind.CANCEL_OPERATION,
        name=f'{full_name}/cancel',
        fn=cancel,
        metadata={'type': 'cancel-operation'},
    )
    return BackgroundAction(
        start_action=start_action,
        check_action=check_action,
        cancel_action=cancel_action,
    )
